src/ape/types/coverage.py

```diff
@@ -0,0 +1,596 @@
+        attribs["lines_valid"] = self.lines_valid
+        attribs["line_rate"] = self.line_rate
+
+        return attribs
+
+    def include(self, contract_name: str) -> ContractCoverage:
+        """
+        Ensure a contract is included in the report.
+        """
+
+        for contract in self.contracts:
+            if contract.name == contract_name:
+                return contract
+
+        # Include the contract.
+        contract_cov = ContractCoverage(name=contract_name)
+        self.contracts.append(contract_cov)
+        return contract_cov
+
+
+class CoverageProject(BaseModel):
+    """
+    A project with coverage collected.
+    """
+
+    name: str
+    """
+    The name of the project being covered.
+    """
+
+    sources: List[ContractSourceCoverage] = []
+    """
+    Coverage for each source in the project.
+    """
+
+    @property
+    def lines(self) -> List[CoverageStatement]:
+        """
+        All valid coverage lines from every function in every contract in every source
+        in this project.
+        """
+
+        statements = []
+        for src in self.sources:
+            statements.extend(src.statements)
+
+        return statements
+
+    @property
+    def lines_covered(self) -> int:
+        """
+        The number of lines with a hit count greater than zero from every function
+        in every contract in every source in this this project.
+        """
+
+        count = 0
+        for src in self.sources:
+            count += src.lines_covered
+
+        return count
+
+    @property
+    def lines_valid(self) -> int:
+        """
+        The number of lines valid for coverage.
+        """
+        return len(self.lines)
+
+    @property
+    def miss_count(self) -> int:
+        """
+        The number of lines missed.
+        """
+        return self.lines_valid - self.lines_covered
+
+    @property
+    def line_rate(self) -> float:
+        """
+        The number of lines hit divided by number of lines.
+        """
+        return self.lines_covered / self.lines_valid
+
+    def dict(self, *args, **kwargs) -> dict:
+        attribs = super().dict(*args, **kwargs)
+
+        # Add coverage stats.
+        attribs["lines_covered"] = self.lines_covered
+        attribs["lines_valid"] = self.lines_valid
+        attribs["line_rate"] = self.line_rate
+
+        return attribs
+
+    def include(self, source_id: str) -> ContractSourceCoverage:
+        for src in self.sources:
+            if src.source_id == source_id:
+                return src
+
+        # Make sure is included.
+        source_cov = ContractSourceCoverage(source_id=source_id)
+        self.sources.append(source_cov)
+        return source_cov
+
+
+class CoverageReport(BaseModel):
+    """
+    Coverage report schema inspired from coverage.py.
+    """
+
+    timestamp: int
+    """
+    The timestamp the report was generated.
+    """
+
+    projects: List[CoverageProject] = []
```

Comment:
Is this so we can track the coverage of dependencies?

Reply:
that, or multi-project test suites, which may not be a thing but could be.

i am inspired by the schema Python uses for coverage...  at this level in Python, this would be a package. However, for ape project land, I was thinking project is the closest resemblance.